src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,50 +1,194 @@
+            after.prev = newNode;
+            before.next = newNode;
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list.size() == 0) {
+            return false;
+        }
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> setNode = findByIndex(index);
+        T oldValue = setNode.element;
+        setNode.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removeNode = findByIndex(index);
+        return removeHelper(removeNode);
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = first;
+        Node<T> removeNode = null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(currentNode.element, object)) {
+                removeNode = currentNode;
+                break;
+            }
+            currentNode = currentNode.next;
+        }
+
+        if (removeNode == null) {
+            return false;
+        }
+        removeHelper(removeNode);
+        return true;
+    }
+
+    private T removeHelper(Node<T> removeNode) {
+        Node<T> before = removeNode.prev;
+        Node<T> after = removeNode.next;
+        final T removeElement = removeNode.element;
+
+        if (before == null) {
+            first = after;
+        } else {
+            before.next = after;
+            removeNode.prev = null;
+        }
+
+        if (after == null) {
+            last = before;
+        } else {
+            after.prev = before;
+            removeNode.next = null;
+        }
+        removeNode.element = null;
+        size--;
+        return removeElement;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findByIndex(int index) {
+        indexOutOfBoundsExceptionCheck(index);
+        if (index == 0) {
+            return first;
+        } else if (index == size - 1) {
+            return last;
+        }
+        return findNode(index);
+    }
+
+    private Node<T> findNode(int index) {
+        int variableForCatchIndexFirst = 0;
+        int variableForCatchIndexLast = size - 1;
+        Node<T> findNode;
+        if (index <= size / 2) {
+            findNode = first;
+            while (variableForCatchIndexFirst <= size / 2) {
+                findNode = findNode.next;
+                if (++variableForCatchIndexFirst == index) {
+                    return findNode;
+                }
```

Comment:
This logic can help to remove checking for first and last element in the method findByIndex(). And we can combine these methods into one
```suggestion
                if (variableForCatchIndexFirst == index) {
                    return findNode;
                }
                findNode = findNode.next;
                variableForCatchIndexFirst++;
```

Reply:
done